File: core/src/main/java/com/core/viewmodel/AbstractViewModel.java
package com.core.viewmodel;

import androidx.lifecycle.ViewModel;

import io.reactivex.disposables.CompositeDisposable;
import io.reactivex.disposables.Disposable;

/**
 * @author volkanhotur
 */

public abstract class AbstractViewModel extends ViewModel implements ViewModelProtocol {

    private CompositeDisposable compositeDisposable;

    public AbstractViewModel(CompositeDisposable compositeDisposable) {
        this.compositeDisposable = compositeDisposable;
    }

    @Override
    public void addDisposable(Disposable disposable) {
        this.compositeDisposable.add(disposable);
    }

    @Override
    public void removeDisposable(Disposable disposable) {
        this.compositeDisposable.remove(disposable);
    }

    @Override
    public void dispose() {
        if (!compositeDisposable.isDisposed()) {
            compositeDisposable.dispose();
        }
    }

    @Override
    public void clear() {
        compositeDisposable.clear();
    }

    @Override
    protected void onCleared() {
        super.onCleared();
        this.dispose();
    }
}

File: core/src/main/java/com/core/viewmodel/AbstractViewModel.kt
package com.core.viewmodel

import androidx.lifecycle.ViewModel
import io.reactivex.disposables.CompositeDisposable
import io.reactivex.disposables.Disposable

/**
 * @author volkanhotur
 */
abstract class AbstractViewModel(private val compositeDisposable: CompositeDisposable?) : ViewModel(), ViewModelProtocol {

    override fun addDisposable(disposable: Disposable?) {
        disposable?.let {
            compositeDisposable?.add(it)
        }
    }

    override fun removeDisposable(disposable: Disposable?) {
        disposable?.let {
            compositeDisposable?.remove(it)
        }
    }

    override fun dispose() {
        compositeDisposable?.let {
            if (!it.isDisposed) {
                it.dispose()
            }
        }
    }

    override fun clear() {
        compositeDisposable?.clear()
    }

    override fun onCleared() {
        super.onCleared()
        dispose()
    }
}